### ISSUE TITLE
[Large Request]Wii U build that features ability to load and link 2 GBAs at once, one on TV, one on Wii U Gamepad

### ISSUE THREAD

**Issue by WiiUFunnn** [mgba-emu/mgba#1776]
So I'm curious if a Wii U build could be made instead of just the Wii version we do have.  I ask so that maybe the full capablities of the Wii U can be recognized.  Obviously a modded Wii U will have no problem playing the Wii version, but if a version for the Wii U were made, ideally it's features could be taken advantage of, and if the system is powerful enough, then maybe it could emulated 2 GBAs at once, have them capable of linking up, and have one GBA display on the TV and the other on the Wii U gamepad.  This alone would inject so much new life into the Wii U scene, as it would open up a lot of great multiplayer Gameboy and GBA experiences that go largely unrealized today due to not many people carrying around gameboys and link cables anymore.

**Comment by endrift** [comment]
Neat idea, but I highly doubt I'll ever do it. I'll leave this open regardless though.

**Comment by Dimensional** [comment]
The biggest issue I see with this idea that would have to rectified
would be emulating 2 GBAs at once. It would put a lot more strain on the
CPU. First step at the moment is getting a native Wii U build before
considering running dual sessions for single-system multiplayer. Once
that's built, and can switch between TV and pad for displaying, then
research can be done into the multiplayer idea.

**Comment by WiiUFunnn** [comment]
True, I'm not sure if the Wii U could handle emulating 2 GBAs at once, at least at decent enough speeds.  The Wii U can emulate 2 GBCs at the same time with the SameBoy core in Retroarch, but even that doesn't run at full speed when both games are full color (but that's also in Retroarch, which isn't exactly being tailor made for the Wii U, so I theorize the Wii U could at least get near full speed with 2 GBCs being emulated, as we know it can already handle that decently as it is.

**Comment by Extrems** [comment]
It certainly could. Even on just one core.

**Comment by Dimensional** [comment]
That's true about the 2 GBCs. RetroArch does add a little bit of overhead, but not enough to make it's removal noticable. And those are GBC games, not GBA. That alone has a more advanced processor with more complex coding. If you want to know if your emulator will run even just a single GBA game without issues, Golden Sun is always the GoTo for benchmarking. :)

Looking over the specs for the Wii U, it has 3 CPU cores and a single GPU. If the emulator can run smoothly on a single core, in theory it could be coded for a second 'child' instance running on a second core for the multiplayer, and then the GPU and 3rd CPU core might be usable for offloading graphics processing.

Again, that's all theory. The architectural differences between the WiiU (PPC 1.243 GHz) and the GBA (ARM7tdmi 16.78MHz) alone would be a problem to iron out.

**Comment by Dimensional** [comment]
If there ends up being native Wii U builds of mGBA, I can help test them as I just got a Wii U and am working on getting homebrew on it. I'm going to be using a 128GB mSD card in the system, so there would be plenty of room for logs and activities if a debug build were used.

**Comment by Dimensional** [comment]
I didn't realize until just recently that there is a WiiU build of the LibRetro Core for mGBA. I never noticed for semi-obvious reasons.
